### ISSUE TITLE
Should path:eval-symlinks behavior be changed for invalid paths?

### ISSUE THREAD

**Issue by krader1961** [elves/elvish#1236]
**TL;DR:**

The `path:eval-symlinks` command should just return the input path when it does not exist rather than throwing a "no such file or directory" exception. To be determined is what to do about other errors returned by `os.Lstat`. Should all errors result in returning the original path or should only the "not exists" error be treated specially? It's not clear what other error conditions `os.Lstat` can return or how to distinguish between the various possibilities.

Note that the external `realpath` command, that `path:eval-symlinks` is intended to mimic, exits with a zero status when the path is invalid and simply regurgitates the input path.

**The long version:**

I noticed @xiaq fixed a bug in the `path:is-dir` and `path:is-regular` commands I introduced. Specifically, the documentation stated that a symlink (in the final component) would always result in those commands returning false. If you want symlinks to be ignored you first need to apply `path:eval-symlinks`. My implementation used `os.Stat` when it should have used `os.Lstat`. However, @xiaq's fix caused me to notice something quite annoying about `path:eval-symlinks`.

My *~/.elvish/rc.elv* script used to contain this statement:

```
if (not (path:is-dir $p)) {
```

That `path:is-dir` outputs `$false` if the path does not refer to a directory; even if it is a symlink to a directory. I made the naive update:

```
if (not (path:is-dir (path:resolve-symlinks $p)) {
```

That results in `path:resolve` throwing an exception if the path does not exist. I instead have to write this:

```
if (not (path:is-dir (try { path:eval-symlinks $p } except { put $p }))) {
```

That seems suboptimal. Note that the full context is this block of code I now have near the bottom of my *~/.elvish/rc.elv* script:

```
# Warn about any non-existent directories in $E:PATH.
each [p]{
  if (not (path:is-dir (try { path:eval-symlinks $p } except { put $p }))) {
    echo (styled "Warning: directory "$p" in $paths does not exist." red)
  }
} $paths
```

Note that the external `realpath` command, that `path:eval-symlinks` mimics, exits with a zero status when the path is invalid and simply regurgitates the input path.

**Comment by hanche** [comment]
As @krader1961 has mentioned in chat, this issue and #1240 are probably related.

**Comment by xiaq** [comment]
Hmm, so the motivation is to make `path:eval-symlinks` easier to chain with `path:is-dir` and similar commands. A valid use case, but the mechanism is non-obvious and almost hacky.

In particular, there is no natural reason that `path:eval-symlinks` should return the original path for broken symlinks; it can just return any invalid path for this to work; `""`, `"\0"` or `$original-path/trailing/elements` would all work and allows the caller to still distinguish broken symlinks from non-symlinks. That said, I don't think these are good choices either - the current behavior of raising an exception is the least surprising one for any user of `path:eval-symlinks`.

Returning to the original use case, IMO it's best to just do it the plain way and support a `&follow-symlinks` option to `path:is-dir`, `path:is-regular`, etc.

**Comment by krader1961** [comment]
@xiaq, This is an example where my multi-decade use of commands like `realpath` unduly influenced my expectations of Elvish behavior. While the behavior of `realpath` can be justified in the context of a POSIX shell I now agree that is not appropriate for Elvish as expressed by its `path:eval-symlinks` command.

P.S., As a practical matter implementing the behavior of the legacy external `realpath` command is extremely difficult given the Go stdlib behavior. I expected to publish an updated P.R. that made the ignoring an invalid symlink optional several days ago. That didn't happen because I couldn't see a straightforward way to implement the legacy `realpath` behavior. The best I could do was complicated and fragile. I was hoping that "sleeping on the question" would allow my subconscious to yield a good solution. That didn't happen so I'm closing this issue. I agree that simply adding an option to the `path:is-*` commands to resolve symlinks is preferable.
